### ISSUE TITLE
Main thread blocked due to disk operation by UserSettingsManager

### ISSUE THREAD

**Issue by AkshayChordiya** [facebook/facebook-android-sdk#743]
## Checklist

- [x] I've updated to the latest released version of the SDK
- [x] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-android-sdk/issues)
- [x] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [x] I've read the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] This issue is not security related and can safely be disclosed publicly on GitHub

## Environment

- Java Version: `1.8`
- Android SDK Version: `29`
- Android Gradle Plugin version:  `3.6.0`

## Expected Results

The main thread should be blocked and the disk operation should be performed on background thread.

## Actual Results

The main thread is blocked due to an IO operation on main thread hence causing an ANR.

**Stacktrace**

```
main" tid=1 Native Performing disk I/O
"main" prio=5 tid=1 Native
  | group="main" sCount=1 dsCount=0 obj=0x759c2000 self=0xb7558ba0
  | sysTid=6718 nice=0 cgrp=apps sched=0/0 handle=0xb6fa4ec8
  | state=D schedstat=( 839039153 181775671 1422 ) utm=61 stm=22 core=2 HZ=100
  | stack=0xbe617000-0xbe619000 stackSize=8MB
  | held mutexes=
  #00  pc 0000000000038e24  /system/lib/libc.so (fsync+12)
  #01  pc 000000000001e29b  /system/lib/libjavacore.so (???)
  #02  pc 00000000000723ff  /system/framework/arm/boot.oat (Java_libcore_io_Posix_fsync__Ljava_io_FileDescriptor_2+102)
  at libcore.io.Posix.fsync (Posix.java)
  at libcore.io.BlockGuardOs.fsync (BlockGuardOs.java:143)
  at java.io.FileDescriptor.sync (FileDescriptor.java:74)
  at android.os.FileUtils.sync (FileUtils.java:152)
  at android.app.SharedPreferencesImpl.writeToFile (SharedPreferencesImpl.java:598)
  at android.app.SharedPreferencesImpl.access$800 (SharedPreferencesImpl.java:51)
  at android.app.SharedPreferencesImpl$2.run (SharedPreferencesImpl.java:512)
- locked <@addr=0x22c000d8> (a java.lang.Object)
  at android.app.SharedPreferencesImpl.enqueueDiskWrite (SharedPreferencesImpl.java:533)
  at android.app.SharedPreferencesImpl.access$100 (SharedPreferencesImpl.java:51)
  at android.app.SharedPreferencesImpl$EditorImpl.commit (SharedPreferencesImpl.java:455)
  at **com.facebook.UserSettingsManager.writeSettingToCache** (UserSettingsManager.java:193)
  at com.facebook.UserSettingsManager.setAutoLogAppEventsEnabled (UserSettingsManager.java:335)
  at com.facebook.FacebookSdk.setAutoLogAppEventsEnabled (FacebookSdk.java:897)
  at com.biowink.clue.analytics.wrappers.FacebookAnalyticsWrapper$Companion.setupPersonalDataSharing$clue_android_app_productionRelease (FacebookAnalyticsWrapper.java:58)
  at com.biowink.clue.analytics.wrappers.FacebookAnalyticsWrapper.onApplicationCreated (FacebookAnalyticsWrapper.java:31)
  at com.biowink.clue.analytics.wrappers.AnalyticsWrapperList.onApplicationCreated (AnalyticsWrapperList.java:11)
  at com.biowink.clue.analytics.AnalyticsManager.onApplicationCreated (AnalyticsManager.java:42)
  at com.biowink.clue.ClueApplication.setupAnalytics (ClueApplication.java:98)
  at com.biowink.clue.ClueApplication.onCreate (ClueApplication.java:73)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1018)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:4934)
  at android.app.ActivityThread.access$1600 (ActivityThread.java:168)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1454)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:135)
  at android.app.ActivityThread.main (ActivityThread.java:5753)
  at java.lang.reflect.Method.invoke! (Native method)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1405)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1200)
```

## Steps to Reproduce

We have been seeing this ANR mostly on Samsung devices with Android 5.0, 5.1 and 6.0. Just opening the app and initializing the facebook SDK causes this ANR since the main thread gets blocked.

**Comment by grebulon** [comment]
Also seeing similar ANRs (same stack), also mostly on older Samsung devices
I'm using facebook-android-sdk:7.0.0

**Comment by kaushikmahato** [comment]
Same, I'm using sdk 5.15.2 and still getting these ANRs mostly on Samsung devices

**Comment by grebulon** [comment]
It's not fixed. There are still deadlocks.

**Comment by kaitian521** [comment]
met same problems

**Comment by kaitian521** [comment]
why not use apply instead of commit when operating sharedpreference?

**Comment by linmx0130** [comment]
Hi @AkshayChordiya, thanks for raising this issue! The bug fix is committed as 77add99aea06d7da03bbf0c33c2db778b9061974 and it will be released in v12.0. Thank you for your patience!
